### PR TITLE
fix silent correctness bug with channels_last usage of upsample cuda kernels

### PR DESCRIPTION
**Pull request by bdhirsh**
Fixes https://github.com/pytorch/pytorch/issues/54590

After the porting the upsample operators to be structured, they now forward memory_format information to the output. This is a problem for the cuda kernels, which are not implemented to deal with `torch.channels_last` memory format. After the structured port, the kernels now expect the output to already be made contiguous if it was not explicitly passed in. The cuda kernel is written to only handle "normal" contiguous tensors, and not channels_last contiguous tensors. The updated operators are:
* upsample_nearest2d
* upsample_nearest3d

It looks like upsample_bilinear2d and upsample_trilinear3d don't have this issue, so I didn't change those kernels.

This fix just allocates a temporary, contiguous output tensor when that happens, writes the results to the temporary and copies the results back to the output tensor.

Also added tests for upsample_nearest2d/3d to explicitly check that cuda and cpu results are the same. Those tests need to have input tensors with a mix of values and a large enough dimension size to distinguish between contiguous and channels_last behavior.

Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#54744 fix silent correctness bug with channels_last usage of upsample cuda kernels**

Differential Revision: [D27351393](https://our.internmc.facebook.com/intern/diff/D27351393)

### CODE REVIEW

File: aten/src/ATen/native/cuda/UpSampleBilinear2d.cu [hunk 1]
```diff
@@ -179,7 +184,7 @@ static void upsample_bilinear2d_out_cuda_template(
         using accscalar_t = at::acc_type<scalar_t, true>;
 
         auto idata = input.packed_accessor64<scalar_t, 4>();
-        auto odata = output.packed_accessor64<scalar_t, 4>();
+        auto odata = output_c.packed_accessor64<scalar_t, 4>();
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/UpSampleBilinear2d.cu:187]: It looks like `packed_accessor` correctly handles the strides of the tensor, and bilinear/trilinear interpolation produce correct results even without this fix, so maybe undo these changes?
Difference of 1e-4 looks kinda large, but maybe that's and in any case it's not changed by this

**Review by ngimel — COMMENTED** [review]
Cool, please add a test!

**Comment by bdhirsh** [comment]
@ngimel Took out the bilinear/trilinear changes and added explicit tests. I confirmed that the tests fail on master and work after this change, so the inputs are "complicated enough" to hit this case.

File: test/test_nn.py [hunk 2]
```diff
@@ -9247,6 +9247,13 @@ def test_upsamplingNearest2d(self):
             gradcheck(lambda x: F.interpolate(x, 4, mode='nearest'), [input])
             gradgradcheck(lambda x: F.interpolate(x, 4, mode='nearest'), [input])
 
+            # Assert that cpu and cuda handle channels_last memory format in the same way
+            # https://github.com/pytorch/pytorch/issues/54590
+            a = torch.arange(48).reshape(2, 2, 3, 4).contiguous(memory_format=torch.channels_last).float()
+            out_cpu = torch.nn.functional.interpolate(a, scale_factor=2, mode='nearest')
+            out_cuda = torch.nn.functional.interpolate(a.to('cuda'), scale_factor=2, mode='nearest')
```
Comment (ngimel) [inline test/test_nn.py:9254]: this will fail if cuda is not available?

**Review by ngimel — COMMENTED** [review]
You are fixing (and testing) only forward, I suspect the same thing is happening in backward, can you please add a test and fix if necessary?

**Review by ezyang — APPROVED** [review]
Perf is gonna be worse but I think this is for the best.

**Comment by ezyang** [comment]
Plz address ngimel's comments

**Comment by bdhirsh** [comment]
@ngimel comments addressed, thanks. For the test, I just added a `cuda.is_available()` check.

For the backwards calls: `upsample_nearest2d/3d_backward` don't currently have channels_last logic in either cpu or cuda, so their meta functions currently don't forward memory_format information. (actually, none of the backwards ops do- I only added the memory_format-forwarding logic to the forward kernels in this [PR](https://github.com/pytorch/pytorch/pull/53535)). A test is definitely necessary though- I added calls to gradcheck/gradgradcheck in the original test. Let me know if that makes sense!

**Comment by ezyang** [comment]
This is fine for now, though I suspect that backwards should also forward memory layout (and people just didn't get around to doing it)
